### PR TITLE
Bug#877 Reconnecting after game over

### LINKED ISSUE #877: [Bug]: Reconnecting after Game over leads to rematch disruption
### Bug Summary

Refreshing after game is over leads to loss of dialog and can't rematch

### Detailed Description

Currently there is no logic in place to continue if you have to refresh after the `GameOverDialog.vue` is displayed, this leads to the inability to continue playing if you want to enter a rematch.

### Current Behavior

Error

### Expected Behavior

After refreshing, the rematch or Game Over dialogs should be visible

### PR DESCRIPTION
**Pull request by Haviles04**
## Issue number

Relevant [issue number](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- Resolves #877 

## Please check the following

- [x] Do the tests still pass? (see [Run the Tests](https://github.com/cuttle-cards/cuttle/blob/main/docs/setup-development.md#run-the-tests))
- [x] Is the code formatted properly? (see [Linting (Formatting)](https://github.com/cuttle-cards/cuttle/blob/main/docs/setup-development.md#linting-formatting))
- For New Features:
  - [x] Have tests been added to cover any new features or fixes?
  - [ ] Has the documentation been updated accordingly?

## Please describe additional details for testing this change

- Added logic to API endpoints to save the Full game at game over into a `lastEvent` variable
- Added logic to `mustReauthenticate` to pull from the `lastEvent.game` if game is over
- Added test to make sure refresh correctly recovers state through lastEvent

### CODE REVIEW

File: api/controllers/game/rematch.js [hunk 1]
```diff
@@ -21,7 +21,14 @@ module.exports = async function (req, res) {
 
     // Determine whether to start new game
     const oldPNum = game.p0 === userId ? 0 : 1;
-    const gameUpdates = { [`p${oldPNum}Rematch`]: rematch };
+    const gameUpdates = {
+      [`p${oldPNum}Rematch`]: rematch,
+      lastEvent: {
+        change: 'rematch',
+        game: { ...game.lastEvent.game, [`p${oldPNum}Rematch`]: rematch },
```
Comment (itsalaidbacklife) [inline api/controllers/game/rematch.js:28]: I'm wondering if we can DRY up the rematch expression by capturing it in a variable like
```suggestion
    const rematchVal = { [`p${oldPNum}Rematch`]: rematch };
    const gameUpdates = {
      ...rematchVal,
      lastEvent: {
        change: 'rematch',
        game: { ...game.lastEvent.game, ...rematchVal },
```

File: api/controllers/game/resolve.js [hunk 2]
```diff
@@ -358,7 +358,7 @@ module.exports = function (req, res) {
         game: fullGame,
         gameModel,
       });
-
+
```
Comment (itsalaidbacklife) [inline api/controllers/game/resolve.js:361]:
```suggestion

```

File: api/controllers/game/resolve.js [hunk 3]
```diff
@@ -369,6 +369,13 @@ module.exports = function (req, res) {
       });
       // If the game is over, clean it up
       if (victory.gameOver) {
+        await Game.updateOne({ id: fullGame.id }).set({
+          lastEvent: {
+            change: 'winByResolvingOneOff',
```
Comment (itsalaidbacklife) [inline api/controllers/game/resolve.js:374]:
```suggestion
            change: 'resolve',
```

File: api/controllers/game/seven/face-card.js [hunk 4]
```diff
@@ -69,6 +69,13 @@ module.exports = function (req, res) {
       });
       // If the game is over, clean it up
       if (victory.gameOver) {
+        await Game.updateOne({ id: fullGame.id }).set({
+          lastEvent: {
+            change: 'winBySevenFaceCard',
```
Comment (itsalaidbacklife) [inline api/controllers/game/seven/face-card.js:74]:
```suggestion
            change: 'sevenFaceCard',
```

File: api/controllers/game/seven/jack.js [hunk 5]
```diff
@@ -116,6 +116,13 @@ module.exports = function (req, res) {
       });
       // If the game is over, clean it up
       if (victory.gameOver) {
+        await Game.updateOne({ id: fullGame.id }).set({
+          lastEvent: {
+            change: 'winBySevenJack',
```
Comment (itsalaidbacklife) [inline api/controllers/game/seven/jack.js:121]:
```suggestion
            change: 'sevenJack',
```

File: api/controllers/game/seven/points.js [hunk 6]
```diff
@@ -59,6 +59,13 @@ module.exports = function (req, res) {
       });
       // If the game is over, clean it up
       if (victory.gameOver) {
+        await Game.updateOne({ id: fullGame.id }).set({
+          lastEvent: {
+            change: 'winBySevenPoints',
```
Comment (itsalaidbacklife) [inline api/controllers/game/seven/points.js:64]:
```suggestion
            change: 'sevenPoints',
```

File: api/controllers/game/stalemate.js [hunk 7]
```diff
@@ -25,30 +24,38 @@ module.exports = async function (req, res) {
     };
 
     gameUpdates.lastEvent = { change: 'requestStalemate', requestedByPNum: pNum };
-
+
```
Comment (itsalaidbacklife) [inline api/controllers/game/stalemate.js:27]:
```suggestion

```

File: api/controllers/game/stalemate.js [hunk 8]
```diff
@@ -25,30 +24,38 @@ module.exports = async function (req, res) {
...
+          game: game,
+          victory
+        }
+      });      
+      await gameService.clearGame({ userId });
+    } else {
+      await Game.updateOne({ id: gameId }).set(gameUpdates);
     }
-
+
```
Comment (itsalaidbacklife) [inline api/controllers/game/stalemate.js:55]:
```suggestion

```

File: api/controllers/game/stalemate.js [hunk 9]
```diff
@@ -25,30 +24,38 @@ module.exports = async function (req, res) {
...
+      await gameService.clearGame({ userId });
+    } else {
+      await Game.updateOne({ id: gameId }).set(gameUpdates);
     }
-
+    
     Game.publish([game.id], {
-      change: 'requestStalemate',
-      game,
+      change:'requestStalemate',
```
Comment (itsalaidbacklife) [inline api/controllers/game/stalemate.js:57]:
```suggestion
      change: 'requestStalemate',
```

File: api/controllers/game/stalemate.js [hunk 10]
```diff
@@ -25,30 +24,38 @@ module.exports = async function (req, res) {
...
+      
+      await Game.updateOne({ id: gameId }).set(gameUpdates);
+      game = await gameService.populateGame({ gameId }); 
+      
+      victory.gameOver = true;
       victory.currentMatch = await sails.helpers.addGameToMatch(game);
+      
+      await Game.updateOne({ id: gameId }).set({
+        lastEvent: {
+          change: 'stalemate',
```
Comment (itsalaidbacklife) [inline api/controllers/game/stalemate.js:46]:
```suggestion
          change: 'reqeuestStalemate',
```

File: src/stores/auth.js [hunk 11]
```diff
@@ -117,9 +117,14 @@ export const useAuthStore = defineStore('auth', {
           const { gameId } = route.params;
           gameStore.requestSpectate(Number(gameId));
         }
-
+
```
Comment (itsalaidbacklife) [inline src/stores/auth.js:120]:
```suggestion

```

File: src/stores/auth.js [hunk 12]
```diff
@@ -117,9 +117,14 @@ export const useAuthStore = defineStore('auth', {
...
         }
-
+        
         if (gameId && (isGame || isLobby)) {
           await this.requestReauthenticate({ username }).then(({ game }) => {
+            if (game.status === GameStatus.FINISHED) {
+              gameStore.updateGame(game.lastEvent.game);
+              gameStore.setGameOver(game.lastEvent.victory);
+              return;
+            }
```
Comment (itsalaidbacklife) [inline src/stores/auth.js:127]: Okay I see now that we're not emitting the event for the gameover case, we're just using the returned game from the response to the reLogin request. That has the nice property of not bothering the other user who is likely still connected and doesn't need the socket event.

The only concern I have is whether there is cleanup logic in the socket event handlers that's missed by calling updateGame and setGameOver w/o firing the actual socket event handlers themself e.g. other variables that need resetting.

File: tests/e2e/specs/in-game/reconnecting.spec.js [hunk 13]
```diff
@@ -699,3 +699,219 @@ describe('Reauthenticating in game', () => {
...
 });
+
+describe('Reconnecting after game is over', () => {
+  beforeEach(() => {
+    cy.setupGameAsP0();
+  });
+  
+  it('Dialogs persist after refreshing when game is over by conceded and opponent request rematch', () => {
+    cy.concedeOpponent();
+    cy.get('[data-cy=game-over-dialog]').should('be.visible');
```
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/reconnecting.spec.js:710]: How about throwing another `cy.reload()` here and asserting the dialog shows up before anyone rematches? Just another moment in time that someone could disconnect

File: tests/e2e/specs/in-game/reconnecting.spec.js [hunk 14]
```diff
@@ -699,3 +699,219 @@ describe('Reauthenticating in game', () => {
...
+    cy.reload();
+    cy.get('[data-cy=game-over-dialog]').should('be.visible');
+    cy.get('[data-cy=match-score-counter-wins]').should('contain', 'W: 1');
+    cy.get('[data-cy=lobby-ready-card]').should('exist');
+  });
+
+  it('Dialogs persist after refreshing when game is over by conceded and player request rematch', () => {
+    cy.concedeOpponent();
+    cy.get('[data-cy=game-over-dialog]').should('be.visible');
+    cy.get('[data-cy=gameover-rematch]').click();
```
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/reconnecting.spec.js:725]: Same here, we could reload before clicking rematch just to confirm that it still shows up before rematch is hit

File: tests/e2e/specs/in-game/reconnecting.spec.js [hunk 15]
```diff
@@ -699,3 +699,219 @@ describe('Reauthenticating in game', () => {
...
+    cy.get('[data-cy=lobby-ready-card]').should('exist');
+  });
+  
+  it('Dialogs persist after refreshing when game is over by stalemate', () => {
+    cy.get('#game-menu-activator').click();
+    cy.get('#game-menu').should('be.visible').get('[data-cy=stalemate-initiate]').click();
+    cy.get('#request-gameover-dialog')
+    .should('be.visible')
+    .get('[data-cy=request-gameover-confirm]')
+    .click();
```
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/reconnecting.spec.js:738]:
```suggestion
      .should('be.visible')
      .get('[data-cy=request-gameover-confirm]')
      .click();
```

File: tests/e2e/specs/in-game/reconnecting.spec.js [hunk 16]
```diff
@@ -699,3 +699,219 @@ describe('Reauthenticating in game', () => {
...
+    cy.get('[data-cy=game-over-dialog]').should('be.visible');
+    cy.url().then((url) => {
+      const oldGameId = Number(url.split('/').pop());
+      cy.rematchOpponent({ gameId: oldGameId, rematch: true });
+    });
+    cy.get('[data-cy=lobby-ready-card]').should('be.visible');
+    cy.reload();
+    cy.get('[data-cy=game-over-dialog]').should('be.visible');
+    cy.get('[data-cy=match-score-counter-wins]').should('contain', 'W: 1');
+    cy.get('[data-cy=lobby-ready-card]').should('exist');
```
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/reconnecting.spec.js:719]: We can be more specific for these assertions to check that the card shows the correct player is ready. See `tests/e2e/specs/in-game/game-over/rematch.spec.js` for examples:

```
  cy.get('[data-cy=opponent-rematch-indicator]')
    .find('[data-cy="lobby-card-container"]')
    .should('have.class', 'ready');
```

Comment (itsalaidbacklife) [reply tests/e2e/specs/in-game/reconnecting.spec.js:719]: Then while we're at it we can assert the 'my' card appears and is not ready

File: tests/e2e/specs/in-game/reconnecting.spec.js [hunk 17]
```diff
@@ -699,3 +699,219 @@ describe('Reauthenticating in game', () => {
...
+  });
+
+  it('Dialogs persist after refreshing when game is over by conceded and player request rematch', () => {
+    cy.concedeOpponent();
+    cy.get('[data-cy=game-over-dialog]').should('be.visible');
+    cy.get('[data-cy=gameover-rematch]').click();
+    cy.get('[data-cy=lobby-ready-card]').should('be.visible');
+    cy.reload();
+    cy.get('[data-cy=game-over-dialog]').should('be.visible');
+    cy.get('[data-cy=lobby-ready-card]').should('exist');
```
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/reconnecting.spec.js:729]: Same re: specificity of the assertion on the ready card

File: tests/e2e/specs/in-game/reconnecting.spec.js
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/reconnecting.spec.js:1]: This is great. Out of scope for this PR, but I'd like you and I to put our heads together with @seriouslysean re: how we can reorganize the test files given overlap across features, especially with respect to reconnecting, game over, rematch, and spectate

**Review by itsalaidbacklife — COMMENTED** [review]
Great work! This is going to be a solid fix. Chang requests are in the comments

File: api/controllers/UserController.js [hunk 18]
```diff
@@ -61,16 +61,17 @@ module.exports = {
       if (!loggedIn) {
         await passwordAPI.checkPass(password, user.encryptedPassword);
       }
-
       // Query for game if user is in one
-      const gameId = user.game;
+      const gameId = (user.game || req.session.game) ?? null;
```
[the remark this answers is not in the document]
Comment (Haviles04) [reply api/controllers/UserController.js:65]: I think nullish is better but I'm pretty sure it still needs to be wrapped in parentheses

File: api/controllers/game/pass.js [hunk 19]
```diff
@@ -57,15 +58,23 @@ module.exports = function (req, res) {
...
+            game: fullGame,
+            victory
+          }
+        });
         await gameService.clearGame({ userId: req.session.usr });
       }
       Game.publish([game.id], {
         change: 'pass',
-        game,
+        game: fullGame ?? game,
```
[the remark this answers is not in the document]
Comment (Haviles04) [reply api/controllers/game/pass.js:77]: I think the name is misleading here, I used full game as a new variable for the game that was populated after the winning condition.(the version of game that is stored in the last event)

File: api/controllers/game/rematch.js (hunk 1 above)
Comment (Haviles04) [reply api/controllers/game/rematch.js:28]: Nice call

File: api/controllers/game/pass.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (itsalaidbacklife) [reply api/controllers/game/pass.js:77]: Is the key to this conditional is whether the game is over? If so let's capture whether the game is over in a variable and use that for our conditional checks here and above

File: api/controllers/game/stalemate.js [hunk 20]
```diff
@@ -25,30 +24,36 @@ module.exports = async function (req, res) {
...
+      });      
+      await gameService.clearGame({ userId });
+    } else {
+      await Game.updateOne({ id: gameId }).set(gameUpdates);
     }
-
     Game.publish([game.id], {
       change: 'requestStalemate',
-      game,
+      game: game,
```
Comment (itsalaidbacklife) [inline api/controllers/game/stalemate.js:56]: This is the equivalent shorthand for the previously existing line
```suggestion
      game,
```

File: tests/e2e/specs/in-game/reconnecting.spec.js
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/reconnecting.spec.js:1]: What about if the first player to hit rematch disconnects before the second player hits rematch? Now the game has started without them. 

Can we recover from that currently?

**Review by itsalaidbacklife — COMMENTED** [review]
Adding further comments

File: tests/e2e/specs/in-game/reconnecting.spec.js
Comment (itsalaidbacklife) [reply tests/e2e/specs/in-game/reconnecting.spec.js:1]: @Haviles04 I'm wondering if this is a tricky/complex enough edge case to warrant its own issue?

Comment (Haviles04) [reply tests/e2e/specs/in-game/reconnecting.spec.js:1]: I added lines 95-97 in the auth store to try to combat this, but I think it is definitively enough of an edge case to warrant it's own issue

Comment (itsalaidbacklife) [reply tests/e2e/specs/in-game/reconnecting.spec.js:1]: I'm going to try adding a test case and seeing if it just works b/c it looks like you might have it already

File: api/controllers/UserController.js [hunk 21]
```diff
@@ -61,19 +61,17 @@ module.exports = {
...
+      const gameId = (user.game ?? req.session.game) ?? null;
       const unpopulatedGame = gameId ? await gameService.findGame({ gameId }) : null;
-      // Get populated game if game has started
       const populatedGame =
-        unpopulatedGame && unpopulatedGame.p0Ready && unpopulatedGame.p1Ready
+        unpopulatedGame?.status === GameStatus.STARTED
           ? await gameService.populateGame({ gameId })
           : null;
-
+
```
Comment (itsalaidbacklife) [inline api/controllers/UserController.js:71]:
```suggestion

```

File: api/controllers/UserController.js [hunk 22]
```diff
@@ -61,19 +61,17 @@ module.exports = {
...
-        unpopulatedGame && unpopulatedGame.p0Ready && unpopulatedGame.p1Ready
+        unpopulatedGame?.status === GameStatus.STARTED
           ? await gameService.populateGame({ gameId })
           : null;
-
+      
       req.session.loggedIn = true;
       req.session.usr = user.id;
-
+
```
Comment (itsalaidbacklife) [inline api/controllers/UserController.js:74]:
```suggestion

```

File: api/controllers/UserController.js [hunk 23]
```diff
@@ -87,7 +85,15 @@ module.exports = {
         });
       }
 
-      const game = populatedGame ?? unpopulatedGame;
+      if (unpopulatedGame && unpopulatedGame.lastEvent.victory) {
```
Comment (itsalaidbacklife) [inline api/controllers/UserController.js:88]: We can optionally chain this for brevity
```suggestion
      if (unpopulatedGame?.lastEvent?.victory) {
```

File: tests/e2e/support/commands.js [hunk 24]
```diff
@@ -336,7 +336,7 @@ Cypress.Commands.add('recoverSessionOpponent', (userFixture) => {
 
 Cypress.Commands.add('drawCardOpponent', () => {
   return new Cypress.Promise((resolve, reject) => {
-    io.socket.get('/game/draw', function handleResponse(res, jwres) {
+    io.socket.get('/game/draw', function handleResponse(res, jwres) {
```
Comment (itsalaidbacklife) [inline tests/e2e/support/commands.js:339]:
```suggestion
    io.socket.get('/game/draw', function handleResponse(res, jwres) {
```

File: tests/e2e/specs/in-game/reconnecting.spec.js [hunk 25]
```diff
@@ -699,3 +699,331 @@ describe('Reauthenticating in game', () => {
...
+      });
+    });
+
+    cy.get('[data-opponent-hand-card]').should('have.length', 5);
+    cy.get('#turn-indicator').should('contain', "OPPONENT'S TURN");
+    cy.drawCardOpponent();
+    cy.get('[data-opponent-hand-card]').should('have.length', 6);
+  });
+
+  it('Brings player to the rematch game when player hits rematch, disconnects, then reconnects socket after opponent hit rematch and started game', function () {
```
Comment (itsalaidbacklife) [inline tests/e2e/specs/in-game/reconnecting.spec.js:976]: For consistency, let's use arrow functions unless we're specifically using `this` to grab an alias (which I tried here and couldn't get to work, then forgot to roll back)
```suggestion
  it('Brings player to the rematch game when player hits rematch, disconnects, then reconnects socket after opponent hit rematch and started game', () => {
```

File: api/controllers/UserController.js (hunk 23 above)
Comment (itsalaidbacklife) [reply api/controllers/UserController.js:88]: Also let's put this right below the other unpopulatedGame conditional just to group the similar ones

File: src/stores/auth.js [hunk 26]
```diff
@@ -117,10 +116,12 @@ export const useAuthStore = defineStore('auth', {
           const { gameId } = route.params;
           gameStore.requestSpectate(Number(gameId));
         }
-
         if (gameId && (isGame || isLobby)) {
           await this.requestReauthenticate({ username }).then(({ game }) => {
             gameStore.updateGame(game);
+            if (+router.currentRoute.value.params.gameId !== game.id) {
```
Comment (itsalaidbacklife) [inline src/stores/auth.js:122]: Let's use Number() to cast for consistency as that's how we're doing it elsewhere

File: src/stores/auth.js [hunk 27]
```diff
@@ -117,10 +116,12 @@ export const useAuthStore = defineStore('auth', {
...
+              router.push(`${isGame ? '/game/' : '/lobby/'}${game.id}`);
```
Comment (itsalaidbacklife) [inline src/stores/auth.js:123]: I'm wondering if this check + navigation is redundant with the socket-driven navigation fix. Just curious if the tests pass without it. That said it's a more general catch all for "server thinks you're in a different game than you are looking at; fix it", so I think it's a good idea

**Review by itsalaidbacklife — APPROVED** [review]
Excellent work. This is a tough problem to solve and it came together nicely. Well done

File: src/stores/auth.js (hunk 27 above)
Comment (Haviles04) [reply src/stores/auth.js:123]: The test do not pass without it, you enter the correct game, but the router does not see it.
